Cython/Compiler/ExprNodes.py

```diff
@@ -4964,9 +4965,11 @@ def analyse_types(self, env, getting=True):
                         new_indices.append(value)
 
             elif index.type.is_int or index.type.is_pyobject:
-                if index.type.is_pyobject and not self.warned_untyped_idx:
+                global_scope = env.global_scope()
+                # We warn about typed indices once for each global scope
+                if index.type.is_pyobject and global_scope not in self.warned_untyped_idx_in_scopes:
                     performance_hint(index.pos, "Index should be typed for more efficient access", env)
-                    MemoryViewIndexNode.warned_untyped_idx = True
+                    MemoryViewIndexNode.warned_untyped_idx_in_scopes.add(global_scope)
```

Comment:
I'd prefer using something simpler like `env.qualified_name` (once per scope), or `env.global_scope().qualified_name` (once per module), to avoid the need for a `WeakSet`.

However, using a class variable here means that we'd then disable the warning entirely for the Cython run once it has been shown. In the case of `IPythonMagic`, Cython can stay in memory for very long. Your way works because you *are* using a `WeakSet`, and scopes should get deleted after compilation, but it generally seems the wrong level to keep track of warnings. That should be done explicitly as part of the compilation state somewhere, not globally and be cleaned up by the accident of cleaning up the module state elsewhere.

Weak-refs are a magical solution to a problem that is best avoided entirely.